csharp/src/Ice/MetricsAdmin.cs

```diff
@@ -144,74 +147,41 @@ public void Failed(string exceptionName)
                 }
                 return ((SubMap<S>)m).GetMatching(helper);
             }
+        }
 
-            public void Detach(long lifetime)
-            {
-                lock (_map)
-                {
-                    _object.TotalLifetime += lifetime;
-                    if (--_object.Current == 0)
-                    {
-                        _map.Detached(this);
-                    }
-                }
-            }
-
-            public void Execute(Observer<T>.MetricsUpdate func)
-            {
-                lock (_map)
-                {
-                    func(_object);
-                }
-            }
-
-            public MetricsMap<T> GetMap() => _map;
-
-            internal MetricsFailures? GetFailures()
-            {
-                if (_failures == null)
-                {
-                    return null;
-                }
-                var f = new MetricsFailures();
-                f.Id = _object.Id;
-                f.Failures = new Dictionary<string, int>(_failures);
-                return f;
-            }
+        private readonly Dictionary<string, Regex> _accept;
+        private LinkedList<Entry>? _detachedQueue;
+        private readonly List<string> _groupByAttributes;
+        private readonly List<string> _groupBySeparators;
+        private readonly object _mutex = new object();
+        private readonly Dictionary<string, Entry> _objects = new Dictionary<string, Entry>();
+        private readonly Dictionary<string, string> _properties;
+        private readonly Dictionary<string, Regex> _reject;
+        private readonly int _retain;
+        private readonly Dictionary<string, ISubMapCloneFactory>? _subMaps;
 
-            internal void Attach(MetricsHelper<T> helper)
+        public MetricsFailures[] GetFailures()
+        {
+            lock (_mutex)
             {
-                ++_object.Total;
-                ++_object.Current;
-                helper.InitMetrics(_object);
+                return _objects.Values.Where(entry => entry.Failures != null).Select(
+                    entry => entry.Failures!.Value).ToArray();
```

Comment:
Can we change this to `ToList()` which should be a little more efficient and return an IReadOnlyList instead?